### DISCUSSION TITLE
can not run any python script

### DISCUSSION THREAD

**Discussion by TNT-DG** [windmill-labs/windmill#3320] [Questions]
After updating my self-hosted server to [v1.282.0], all the python script can not be run. I don't know why and how to fix it.

Here is the log in web console:
getWorker: workerId: workerMain.js label: editorWorkerService
Editor.9bI3LOeY.js:409 started client
Editor.9bI3LOeY.js:409 Workspace settings: null
Editor.9bI3LOeY.js:409 Global settings: {}
Editor.9bI3LOeY.js:409 started client
Editor.9bI3LOeY.js:409 Using interpreter executable: /usr/local/bin/ruff
Editor.9bI3LOeY.js:409 Inferred version 0.2.2 for: /usr/local/bin/ruff
Editor.9bI3LOeY.js:409 Found ruff 0.2.2 at /usr/local/bin/ruff
Editor.9bI3LOeY.js:409 Running Ruff with: /usr/local/bin/ruff ['--force-exclude', '--no-cache', '--no-fix', '--quiet', '--output-format', 'json', '-', '--stdin-filename', '/tmp/monaco/iuamkiue0w.py']
Editor.9bI3LOeY.js:409 [Info  - 13:16:13] Pyright language server 1.1.351 starting
Editor.9bI3LOeY.js:409 [Info  - 13:16:13] Server root directory: file:///usr/local/share/.config/yarn/global/node_modules/pyright/dist
Editor.9bI3LOeY.js:409 [Info  - 13:16:13] Starting service instance "windmill"
Editor.9bI3LOeY.js:409 started client
Editor.9bI3LOeY.js:409 [Error - 13:16:14] File or directory "/tmp/monaco/iuamkiue0w.py" does not exist.
Editor.9bI3LOeY.js:409 [Info  - 13:16:14] No source files found.
EditorTheme.D7nUzZ3v.js:16   ERR MainThreadFileSystem#$watch(): failed to stat a resource for file watching (extension: nullExtensionDescription, path: file:///usr/local/lib/python3.11, recursive: true, session: 0.847132592515452): EntryNotFound (FileSystemError): file not found
EditorTheme.D7nUzZ3v.js:16   ERR MainThreadFileSystem#$watch(): failed to stat a resource for file watching (extension: nullExtensionDescription, path: file:///root/.local/share/virtualenvs/-x-v5uFv0/lib/python3.11/site-packages, recursive: true, session: 0.18656932385241443): EntryNotFound (FileSystemError): file not found
AppPreview.Cq7lmGoz.js:18 started 018df871-0d4d-8b3a-cc11-fc94188079bd
Editor.9bI3LOeY.js:409 [Error - 13:16:15] Stopping server timed out
Editor.9bI3LOeY.js:409 Stopping client early because of mismatch
Editor.9bI3LOeY.js:408 WebSocket is already in CLOSING or CLOSED state.

**Comment by rubenfiszel** [comment]
1.282.0 works fine and your browser logs are not useful. Your server/worker logs might be more useful as well as sharing the actual error

**Reply by TNT-DG** [reply]
I tried to find any useful error log in worker-1,2,3 and server-1 but I failed. Is there any suggestion to debug?
![image](https://github.com/windmill-labs/windmill/assets/29853150/1981ebd2-a9df-4ff5-849b-137024805dc1)

**Reply by TNT-DG** [reply]
I also tried to pull another version of  image v1.281.3 and got the same problem.

**Comment by TNT-DG** [comment]
Finally, I reset my instance and reproduce the problem again. It seems to be related to socketio or my usage. After running several times, all the python script execution will be blocked forever. Even if I has restart the docker container.
![image](https://github.com/windmill-labs/windmill/assets/29853150/075c7eb2-75f8-4fe4-9768-7031fb917579)
![image](https://github.com/windmill-labs/windmill/assets/29853150/631f5ee2-2aed-48f8-8310-3905fc302293)

**Comment by Girgitt** [comment]
I was just hit by a similar issue in v1.298.0. python scripts never start to execute - as background runnables or when directly run from script editing window. restarting containers does not change anything, restarting whole docker service also. everything worked fine for weeks. This is not a proper report by any means - just a comment such things happen with windmill. I need to get it back to work and maybe then will have time to describe the issue.

**Reply by rubenfiszel** [reply]
They do not happen randomly. Windmill is used in critical environment such as remote oil rigs and satellites and it never failed :)

Let us know once you find the root cause

**Reply by Girgitt** [reply]
Maybe the issue was related to networking but it is strange it persisted for 30 or so minutes. Then after I opened all python scripts one by one in the the web editor python execution started to work again. This could be totally unrelated to the issue but who knows. 

When issue persisted there was repeated entry in the console output of docker compose: 

windmill_worker-1         | 2024-07-18T13:51:42.269633Z  WARN request: windmill_common::error: Not found: Script not found at name f/careapp/data_access method=GET uri=/api/w/g4e_careapp/scripts/raw/p/f/careapp/data_access.py

But I do not have data_access.py script at all - it is just a sub-level of path used to store other scripts for convenient import as a module 

After re-opening every python script (without deploying any changes) this warning disappeared  and log has no periodic messages, just stays at:
lsp-1                     | INFO:tornado.access:101 GET /ws/diagnostic (172.23.0.3) 0.65ms
lsp-1                     | INFO:__main__:Spawning pylsp subprocess<__main__.DiagnosticLS object at 0x7f549f611c50>
lsp-1                     | INFO:__main__:CLOSING: <__main__.PyrightLS object at 0x7f549f5c7890>
lsp-1                     | Cancel notification for unknown message id "2"
lsp-1                     | INFO:__main__:CLOSING: <__main__.RuffLS object at 0x7f549f5eb0d0>
lsp-1                     | INFO:__main__:CLOSING: <__main__.DiagnosticLS object at 0x7f549f611c50>
db-1                      | 2024-07-18 14:00:22.225 UTC [27] LOG:  checkpoint complete: wrote 5036 buffers (30.7%); 0 WAL file(s) added, 0 removed, 4 recycled; write=269.959 s, sync=0.194 s, total=270.157 s; sync files=62, longest=0.146 s, average=0.004 s; distance=53729 kB, estimate=304886 kB; lsn=34E/E00B968, redo lsn=34E/12AE728
db-1                      | 2024-07-18 14:00:52.256 UTC [27] LOG:  checkpoint starting: time

When issue persisted there were many (10-20 a second) GET requests logged similar to

windmill_worker-2         | 2024-07-18T14:05:27.162197Z  INFO request: windmill_api::tracing_init: response latency=2 status=200 method=GET uri=/api/w/g4e_careapp/scripts/raw/p/f/careapp/redis_hash.py username="z…y" email="z..l" workspace_id="g4e_careapp"
windmill_worker-2         | 2024-07-18T14:05:27.167869Z  WARN request: windmill_common::error: Not found: Script not found at name f/careapp/data_access method=GET uri=/api/w/g4e_careapp/scripts/raw/p/f/careapp/data_access.py username="z..y" email="z..l" workspace_id="g4e_careapp"
windmill_worker-2         | 2024-07-18T14:05:27.167902Z  INFO request: windmill_api::tracing_init: response latency=1 status=404 method=GET uri=/api/w/g4e_careapp/scripts/raw/p/f/careapp/data_access.py username="z..y" email="z..l" workspace_id="g4e_careapp"
windmill_worker-2         | 2024-07-18T14:05:27.171058Z  INFO request: windmill_api::tracing_init: response latency=1 status=200 method=GET uri=/api/w/g4e_careapp/scripts/raw/p/f/careapp/data_access/query_live_values.py username="z..y" email="z..l" workspace_id="g4e_careapp"
windmill_worker-2         | 2024-07-18T14:05:27.176984Z  INFO request: windmill_api::tracing_init: response latency=1 status=200 method=GET uri=/api/w/g4e_careapp/scripts/raw/p/f/careapp/utils.py username="z..y" email="z..l" workspace_id="g4e_careapp"

I guess many requests related to unfinished (frozen or rather never really started) background python runnables.  

Sorry for hijacking the thread but this was the fastest way to provide feedback. I will archive logs which captured the whole event and try to find the moment when the issue started.

**Reply by Girgitt** [reply]
I agree, Windmill is typically rock-solid, so this event came as a real shock. Fortunately everything went back to normal.

**Reply by Girgitt** [reply]
During the event docker was utilizing 4 cores while windmill stayed at 2-5% reported in lazydocker by each container. On this host docker runs only single Windmill instance via docker-compose. Docker daemon logged to system journal: 

msg="got error while decoding json" error="invalid character 'm' looking for beginning of value" retries=1131
Jul 18 15:43:49 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-18T15:43:48.980901675+02:00" level=warning msg="got error while decoding json" error="invalid character 'r' looking for beginning of value" retries=1132
Jul 18 15:43:49 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-18T15:43:48.980917104+02:00" level=warning msg="got error while decoding json" error="invalid character ':' looking for beginning of value" retries=1133
Jul 18 15:43:49 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-18T15:43:48.980931992+02:00" level=warning msg="got error while decoding json" error="invalid character 'm' after array element" retries=1134
Jul 18 15:43:49 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-18T15:43:48.980946449+02:00" level=warning msg="got error while decoding json" error="json: cannot unmarshal number into Go value of type jsonlog.JSONLog" retries=1135
Jul 18 15:43:49 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-18T15:43:48.980959363+02:00" level=warning msg="got error while decoding json" error="invalid character 'm' looking for beginning of value" retries=1136
Jul 18 15:43:49 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-18T15:43:48.980973740+02:00" level=warning msg="got error while decoding json" error="invalid character 'r' looking for beginning of value" retries=1137
Jul 18 15:43:49 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-18T15:43:48.980988287+02:00" level=warning msg="got error while decoding json" error="invalid character 'g' looking for beginning of value" retries=1138
Jul 18 15:43:49 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-18T15:43:48.981003716+02:00" level=warning msg="got error while decoding json" error="json: cannot unmarshal number into Go value of type jsonlog.JSONLog" retries=1139
Jul 18 15:43:49 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-18T15:43:48.981017051+02:00" level=warning msg="got error while decoding json" error="invalid character 'b' looking for beginning of value" retries=1140
Jul 18 15:43:49 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-18T15:43:48.981030927+02:00" level=warning msg="got error while decoding json" error="invalid character 'b' looking for beginning of value" retries=1141
Jul 18 15:43:49 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-18T15:43:48.981045585+02:00" level=warning msg="got error while decoding json" error="invalid character 'm' looking for beginning of value" retries=1142

Then nothing for the last ~hour when issue stopped. 

A quick search suggests (https://github.com/docker/for-linux/issues/1350) broken log file of one of containers.

**Reply by Girgitt** [reply]
Today there was a similar short outage and again docker logged issues with parsing logs: 

Jul 18 21:16:13 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-18T21:16:13.128824148+02:00" level=warning msg="got error while decoding json" error="invalid character 'b' looking for beginning of value" retries=1141
Jul 18 21:16:13 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-18T21:16:13.128839707+02:00" level=warning msg="got error while decoding json" error="invalid character 'm' looking for beginning of value" retries=1142
Jul 19 11:03:39 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-19T11:03:39.567470121+02:00" level=warning msg="got error while decoding json" error="invalid character 'l' after object key:value pair" retries=0
..
Jul 19 11:03:52 app_main_1.bmscare.cloud dockerd[2210886]: time="2024-07-19T11:03:52.446803684+02:00" level=warning msg="got error while decoding json" error="invalid character 'm' looking for beginning of value" retries=1142


For about 13 seconds - but journal also shows similar case around 21:16 yesterday so issue happens every couple hours

**Reply by Girgitt** [reply]
for now I changed json-file logger to syslog and see if it helps

**Reply by Girgitt** [reply]
No docker hiccups for the last 3 days since logging changed to syslog. Windmill runs smoothly all the time.

**Reply by Girgitt** [reply]
I can confirm in my case the issue with python background jobs' execution was with Docker logging and not with Windmill. 

The json-file driver occasionally fails. 

The default syslog TCP can also fail when the syslog server fails. This happened to me once but with promtail forwarder being the syslog TCP target for containers. 

When a syslog server is unresponsive, and syslog TCP is used, workers can suddenly stop executing jobs and Windmill's App debug window shows long running jobs ending with error similar to "workers tagged with python are not available"; or maybe this error appears after restart attempt and before restart the problematic python jobs end with a timeout.

If you try to start workers with unresponsive syslog server, containers defined with docker-compose stay in "created" state and only dockerd's log mentions syslog connection failure which is not obvious reason for containers not being able to start.  

A proper fix was to switch to syslog UDP which doesn't care about the syslog server (but at least containers start normally and there should be no logging-related issues freezing worker containers).
